tests/always/test_project_structure.py

```diff
@@ -30,68 +30,79 @@
 )
 
 
-class TestProjectStructure(unittest.TestCase):
+class TestProjectStructure:
```

Comment:
I thought `TestProjectStructure` should be re-worked, I miss the main idea of this test so it would be nice if someone could share the idea about this tests

---

`test_deprecated_packages` - I thought this test not properly worked since actual files replaced by PEP-562 getattr - https://github.com/apache/airflow/pull/26153

---

`test_providers_modules_should_have_tests` - This test broken for a while, generators wrapped into generators and when it actually transformed to set this line broke `expected_test_files` calculation and it always `set([])`
https://github.com/apache/airflow/blob/e377e869da9f0e42ac1e0a615347cf7cd6565d54/tests/always/test_project_structure.py#L87

Rather than raise an error in this test, I add print warning instead of until we fix or rewrite this test logic

Reply:
Most of this module come from our past renaming we've doe in [AIP-21](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-21%3A+Changes+in+import+paths)  - and some of them can be removed or simplified or moved elsewhere as they never caught up as "common provider" approach:

* test_deprecated_packages -> we should indeed remove, they are no problem any more
* Test...ProviderProjectStructure - we should remove them - they bring no value and having then for few providers in here makes littel sense
* ExampleCoverageTest -  similarly
* AssetsCoverageTest -> I think this is largely used by Google team. Naybe @bhirsz might tell more about it (though in any case I thik if it is useful for Google it should go to google provider tests.
* test_providers_modules_should_have_tests -> is the only one I find useful from those for now but if it's broken, it might be we are far-off and fixing it would be problematic.

Reply:
`test_providers_modules_should_have_tests` after changes by this PR shows this warning

```console
Detect missing tests in providers module.

Modules Files: 566, Current Test Files: 583, Missing Tests Files: 97.

tests/providers/amazon/aws/hooks/test_dms.py
tests/providers/amazon/aws/links/test_base_aws.py
tests/providers/amazon/aws/links/test_batch.py
tests/providers/amazon/aws/links/test_emr.py
tests/providers/amazon/aws/links/test_logs.py
tests/providers/amazon/aws/operators/test_aws_lambda.py
tests/providers/amazon/aws/operators/test_dms.py
tests/providers/amazon/aws/operators/test_emr.py
tests/providers/amazon/aws/operators/test_s3.py
tests/providers/amazon/aws/operators/test_sagemaker.py
tests/providers/amazon/aws/sensors/test_dms.py
tests/providers/amazon/aws/sensors/test_ecs.py
tests/providers/amazon/aws/sensors/test_emr.py
tests/providers/amazon/aws/sensors/test_s3.py
tests/providers/amazon/aws/sensors/test_sagemaker.py
tests/providers/amazon/aws/test_exceptions.py
tests/providers/amazon/aws/utils/test_rds.py
tests/providers/amazon/aws/waiters/test_base_waiter.py
tests/providers/apache/cassandra/hooks/test_cassandra.py
tests/providers/apache/druid/operators/test_druid_check.py
tests/providers/cncf/kubernetes/backcompat/test_backwards_compat_converters.py
tests/providers/cncf/kubernetes/test_python_kubernetes_script.py
tests/providers/cncf/kubernetes/utils/test_xcom_sidecar.py
tests/providers/databricks/hooks/test_databricks_base.py
tests/providers/databricks/utils/test_databricks.py
tests/providers/dbt/cloud/hooks/test_dbt.py
tests/providers/dbt/cloud/operators/test_dbt.py
tests/providers/dbt/cloud/sensors/test_dbt.py
tests/providers/elasticsearch/log/test_es_json_formatter.py
tests/providers/google/cloud/links/test_base.py
tests/providers/google/cloud/links/test_bigquery.py
tests/providers/google/cloud/links/test_bigquery_dts.py
tests/providers/google/cloud/links/test_bigtable.py
tests/providers/google/cloud/links/test_cloud_build.py
tests/providers/google/cloud/links/test_cloud_functions.py
tests/providers/google/cloud/links/test_cloud_memorystore.py
tests/providers/google/cloud/links/test_cloud_sql.py
tests/providers/google/cloud/links/test_cloud_storage_transfer.py
tests/providers/google/cloud/links/test_cloud_tasks.py
tests/providers/google/cloud/links/test_compute.py
tests/providers/google/cloud/links/test_data_loss_prevention.py
tests/providers/google/cloud/links/test_datacatalog.py
tests/providers/google/cloud/links/test_dataflow.py
tests/providers/google/cloud/links/test_dataform.py
tests/providers/google/cloud/links/test_dataplex.py
tests/providers/google/cloud/links/test_dataprep.py
tests/providers/google/cloud/links/test_dataproc.py
tests/providers/google/cloud/links/test_datastore.py
tests/providers/google/cloud/links/test_kubernetes_engine.py
tests/providers/google/cloud/links/test_life_sciences.py
tests/providers/google/cloud/links/test_mlengine.py
tests/providers/google/cloud/links/test_pubsub.py
tests/providers/google/cloud/links/test_spanner.py
tests/providers/google/cloud/links/test_stackdriver.py
tests/providers/google/cloud/links/test_vertex_ai.py
tests/providers/google/cloud/links/test_workflows.py
tests/providers/google/cloud/operators/vertex_ai/test_auto_ml.py
tests/providers/google/cloud/operators/vertex_ai/test_batch_prediction_job.py
tests/providers/google/cloud/operators/vertex_ai/test_custom_job.py
tests/providers/google/cloud/operators/vertex_ai/test_dataset.py
tests/providers/google/cloud/operators/vertex_ai/test_endpoint_service.py
tests/providers/google/cloud/operators/vertex_ai/test_hyperparameter_tuning_job.py
tests/providers/google/cloud/operators/vertex_ai/test_model_service.py
tests/providers/google/cloud/sensors/test_dataform.py
tests/providers/google/cloud/transfers/test_presto_to_gcs.py
tests/providers/google/cloud/transfers/test_trino_to_gcs.py
tests/providers/google/cloud/triggers/test_cloud_composer.py
tests/providers/google/cloud/triggers/test_dataproc.py
tests/providers/google/cloud/utils/test_bigquery.py
tests/providers/google/cloud/utils/test_bigquery_get_data.py
tests/providers/google/cloud/utils/test_dataform.py
tests/providers/google/common/links/test_storage.py
tests/providers/google/common/test_consts.py
tests/providers/google/test_go_module_utils.py
tests/providers/microsoft/azure/hooks/test_batch.py
tests/providers/microsoft/azure/hooks/test_container_instance.py
tests/providers/microsoft/azure/hooks/test_container_registry.py
tests/providers/microsoft/azure/hooks/test_container_volume.py
tests/providers/microsoft/azure/hooks/test_cosmos.py
tests/providers/microsoft/azure/hooks/test_data_factory.py
tests/providers/microsoft/azure/hooks/test_data_lake.py
tests/providers/microsoft/azure/hooks/test_fileshare.py
tests/providers/microsoft/azure/hooks/test_synapse.py
tests/providers/microsoft/azure/operators/test_adls.py
tests/providers/microsoft/azure/operators/test_batch.py
tests/providers/microsoft/azure/operators/test_container_instances.py
tests/providers/microsoft/azure/operators/test_cosmos.py
tests/providers/microsoft/azure/operators/test_data_factory.py
tests/providers/microsoft/azure/operators/test_synapse.py
tests/providers/microsoft/azure/secrets/test_key_vault.py
tests/providers/microsoft/azure/sensors/test_cosmos.py
tests/providers/microsoft/azure/sensors/test_data_factory.py
tests/providers/mongo/sensors/test_mongo.py
tests/providers/presto/transfers/test_gcs_to_presto.py
tests/providers/redis/operators/test_redis_publish.py
tests/providers/redis/sensors/test_redis_key.py
tests/providers/trino/transfers/test_gcs_to_trino.py'
```

Reply:
- some of tests moved into tests.integrations
- Amazon links tested in one go in same module
- Some of amazon tests split across different test modules, for example: test_s3_bucket.py + test_s3_bucket_tagging.py + test_s3_file_transform.py + test_s3_list.py + test_s3_list_prefixes.py + test_s3_object.py but module is s3.py (merged version)

So in theory there is no problem to fix some issue, add tests.integration as search path and add temporary exclusion for remaining and turn on this test again

Reply:
Yeah. Sounds like a good idea.

Reply:
> * test_deprecated_packages -> we should indeed remove, they are no problem any more

Done

> * Test...ProviderProjectStructure - we should remove them - they bring no value and having then for few providers in here makes littel sense

Done, keep only `ProjectStructureTest` because `ProjectStructureTest` depends on it

> * ExampleCoverageTest -  similarly

Removed

> * AssetsCoverageTest -> I think this is largely used by Google team. Naybe @bhirsz might tell more about it (though in any case I thik if it is useful for Google it should go to google provider tests.

Keep for a while

> * test_providers_modules_should_have_tests -> is the only one I find useful from those for now but if it's broken, it might be we are far-off and fixing it would be problematic.

Will try to fix (make it run again) as follow up